cvmfs/network/sink_mem.h

```diff
@@ -0,0 +1,158 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#ifndef CVMFS_NETWORK_SINK_MEM_H_
+#define CVMFS_NETWORK_SINK_MEM_H_
+
+#include <cassert>
+#include <cstring>
+#include <string>
+
+#include "sink.h"
+#include "util/posix.h"
+#include "util/smalloc.h"
+#include "util/string.h"
+
+namespace cvmfs {
+
+class MemSink : public Sink {
+ public:
+  MemSink() : size_(0), pos_(0), data_(NULL) { is_owner_ = true; }
+  explicit MemSink(size_t size) : size_(size), pos_(0) {
+    data_ = static_cast<char *>(smalloc(size));
+    is_owner_ = true;
+  }
+
+  virtual ~MemSink() { FreeData(); }
+
+  /**
+   * Appends data to the sink
+   * If the sink is too small and
+   * - the sink is the owner of data_: sink size is doubled
+   * - the sink is NOT the owner of data_: fails with -ENOSPC
+   * 
+   * @returns on success: number of bytes written 
+   *          on failure: -errno.
+   */
+  virtual int64_t Write(const void *buf, uint64_t sz) {
+    if (pos_ + sz > size_) {
+      if (is_owner_) {
+        size_t new_size = pos_ + sz < size_ * 2 ? size_ * 2 : pos_ + sz + 1;
+        data_ = static_cast<char *>(srealloc(data_, new_size));
+        size_ = new_size;
+      } else {
+        return -ENOSPC;
+      }
+    }
+
+    memcpy(data_ + pos_, buf, sz);
+    pos_ += sz;
+    return sz;
+  }
+
+  /**
+   * Truncate all written data and start over at position zero.
+   * 
+   * @returns Success = 0
+   *          Failure = -1
+   */
+  virtual int Reset() {
+    int ret = 0;
+    if (is_owner_ && data_) {
```

Comment:
The `free()` call already handles the `NULL` pointer, hence
```suggestion
    if (is_owner_) {
```